### PR TITLE
[FIX] account: inv. report price_average, quantity 

### PR DESCRIPTION
**Pull request by mhabboush**
Description of the issue/feature this PR addresses:
unit conversion from line uom to template uom is incorrect

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by mhabboush** [comment]
@agr-odoo follow-up of partial fix 978012b38abb941c88fe85f1befa2ac659545915

File: doc/cla/individual/mhabboush.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+I hereby agree to the terms of the Odoo Individual Contributor License
+Agreement v1.0.
+
+I declare that I am authorized and able to make this agreement and sign this
+declaration.
+
+Signed,
+
+Mahmoud Habboush mhabboush@users.noreply.github.com https://github.com/mhabboush
```
Comment (william-andre) [inline doc/cla/individual/mhabboush.md:11]: The email needs to be `62886470+mhabboush@users.noreply.github.com`, including the digits and the `+`.

File: addons/account/report/account_invoice_report.py [hunk 2]
```diff
@@ -95,12 +95,12 @@ def _select(self):
                 move.invoice_date_due,
                 uom_template.id                                             AS product_uom_id,
                 template.categ_id                                           AS product_categ_id,
-                line.quantity / NULLIF(COALESCE(uom_line.factor, 1) / COALESCE(uom_template.factor, 1), 0.0) * (CASE WHEN move.move_type IN ('in_invoice','out_refund','in_receipt') THEN -1 ELSE 1 END)
+                line.quantity / NULLIF(COALESCE(uom_line.factor, 1) * COALESCE(uom_template.factor, 1), 0.0) * (CASE WHEN move.move_type IN ('in_invoice','out_refund','in_receipt') THEN -1 ELSE 1 END)
```
Comment (william-andre) [inline addons/account/report/account_invoice_report.py:98]: This one is inside of a parenthesis so this change is wrong.

File: addons/account/report/account_invoice_report.py [hunk 3]
```diff
@@ -95,12 +95,12 @@ def _select(self):
...
                                                                             AS quantity,
                 -line.balance * currency_table.rate                         AS price_subtotal,
                 -COALESCE(line.balance
                    / NULLIF(line.quantity, 0.0)
-                   / NULLIF(COALESCE(uom_line.factor, 1), 0.0)
+                   * NULLIF(COALESCE(uom_line.factor, 1), 0.0)
```
Comment (william-andre) [inline addons/account/report/account_invoice_report.py:103]: What we had before was 
`-line.balance / NULLIF(COALESCE(uom_line.factor, 1) / COALESCE(uom_template.factor, 1), 0.0)`
or more simply `-balance / (uom_factor / template_factor)`
= `-balance / uom_factor * template_factor`

So the `*` sign should be for the line below if I trust the version that was there before the bugfix mentionned.

**Review by william-andre — CHANGES_REQUESTED** [review]

File: addons/account/report/account_invoice_report.py (hunk 3 above)
Comment (agr-odoo) [reply addons/account/report/account_invoice_report.py:103]: I made a test just to be sure of the math

Suppose to have a Product with unit of measure KG and 2 invoice line as follows:

```
  500g  @    1$/g   ->  500$
    1kg @ 1000$/kg  -> 5000$
```

before my patch we used as AveragePrice the formula `balance / (uom_factor / template_factor)`
The report show

|Total|Quantity|AveragePrice|
|------|------------|------------------|
|5,500.00|5.50|2,500.25|

so it seems it was not correct from the beginning
after my patch we have the formula `balance / quantity / (uom_factor / template_factor)`
|Total|Quantity|AveragePrice|
|------|------------|------------------|
|5,500.00|5.50|500.00|

By using the current proposal, we would have the formula 

`-balance / quantity * (uom_factor / template_factor) `

resulting in

|Total|Quantity|AveragePrice|
|------|------------|------------------|
|5,500.00|5.50|1000.00|

which imho **looks correct**

File: addons/account/report/account_invoice_report.py (hunk 2 above)
Comment (mhabboush) [reply addons/account/report/account_invoice_report.py]: Yes I thought the nullif is only for the uom_line.factor, but generally I think only the uom_line.factor should be inside the nullif as other parts should be multiplied

File: addons/account/report/account_invoice_report.py (hunk 3 above)
Comment (william-andre) [reply addons/account/report/account_invoice_report.py:103]: Yeah, don't use multiple division signs on the "same line". The order of execution is not well defined and may vary.
Use a notation like this instead `(a * b * c) / (d * e * f)`, which is a lot easier to understand.

**Comment by william-andre** [comment]
Also, please only keep two commits: the first one for the CLA and the second one for the fix. We don't care about the intermediate fixes in this PR

**Review by mhabboush — COMMENTED** [review]
Exactly, since there were no parenthesis, the (balance / quantity) should be converted using the compute price function

**Comment by mhabboush** [comment]
> Also, please only keep two commits: the first one for the CLA and the second one for the fix. We don't care about the intermediate fixes in this PR

You mean create new PR?

**Comment by william-andre** [comment]
> You mean create new PR?

@mhabboush I meant squasing (or fixing up) your commits and force pushing on this branch. See the `f` command of `git rebase`.
https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

**Comment by william-andre** [comment]
@mhabboush You could have just force pushed on this branch instead of creating a new PR. Everything would have been updated nicely while keeping the history.

**Pull request closed without merging**